server/datastore/mysql/scripts.go

```diff
@@ -569,3 +580,72 @@ func (ds *Datastore) GetHostLockWipeStatus(ctx context.Context, hostID uint, fle
 	}
 	return status, nil
 }
+
+// LockHost will create the script execution request and update host_mdm_actions in a single transaction.
+func (ds *Datastore) LockHostViaScript(ctx context.Context, request *fleet.HostScriptRequestPayload) error {
+	var res *fleet.HostScriptResult
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		var err error
+		res, err = newHostScriptExecutionRequest(ctx, request, tx)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "lock host via script create execution")
+		}
+
+		const stmt = `
+	INSERT INTO host_mdm_actions
+	(
+		host_id,
+		lock_ref,
+		unlock_ref
+	)
+	VALUES (?,?,NULL)
+	ON DUPLICATE KEY UPDATE
+		lock_ref = VALUES(lock_ref),
+		unlock_ref = NULL
```

Comment:
Just so it's able to auto-fix an invalid state, I think I'd set unlock_pin = NULL on duplicate key (and even wipe_ref = NULL since those 3 states are exclusive).

Reply:
Thinking about it more, it's more subtle than that unfortunately - we shouldn't even clear the unlock_ref because this just records the _pending_ request to lock the host, if it was unlocked we still want the unlock reference there (and same in UnlockHost), until the script is executed and then it should update this state... It's quite tricky, we have to track those scripts until resolution (succeeded or failed).